parachain/src/lib.rs

```diff
@@ -91,8 +93,17 @@ pub struct ValidationResult {
 	pub head_data: Vec<u8>,
 }
 
+/// An incoming message.
+#[derive(PartialEq, Eq, Decode)]
+#[cfg_attr(feature = "std", derive(Debug, Encode))]
+pub struct IncomingMessage {
+	/// The source parachain.
+	pub source: u32,
```

Comment:
`ParachainIndex` rather than `u32`?

Reply:
Unfortunately, we can't depend on the `polkadot-primitives` crate here because it introduces unwanted `sr-primitives` and `sr-std` deps into the parachain WASM. Maybe we could define that type here and introduce a reverse dependency.